### PR TITLE
yumpkg.list_pkgs: add parameter for returned attribute selection

### PR DESCRIPTION
**Pull request by moio**
### What does this PR do?

It adds a new optional parameter to `list_pkg`, originally added to the the `zypper` module via PR #42310, to `yumpkg` providing the same functionality and interface to the `yum` package manager.

### What issues does this PR fix or reference?

None known.

### New Behavior

If the new parameter `attr` is specified, more package attributes are returned.

### Tests written?

Yes - existing tests have been adapted as well.

### CODE REVIEW

**Review by isbm — APPROVED** [review]
:+1: Thanks, @moio ! Extra-hug for tests. :wink:

**Comment by damon-atkins** [comment]
There is a lot of work gone into this.
What do we gain with this change vs pkg.info_installed which returns all attributes. Keep in mine  pkg.list_pkgs is used by states and pkg.info_installed is not.

Would it be better for pkg.info_installed  support filtering the attributes, so less attributes are  return.
Just before the data is returned, you could loop though the attributes and remove any not in the attribute filter list.  This would result in all attributes being supported across all platforms and with only a few lines of code.

salt \* pkg.info_installed attr=arch,version,abc

Looks like the _context__['pkg.list_pkgs'] will not work after this change.

**Comment by moio** [comment]
> What do we gain with this change vs pkg.info_installed which returns all attributes

[This was already discussed](https://github.com/saltstack/salt/pull/42310#discussion_r127638393) in the context of the `zypper` module, approved and merged. I see no difference here: this PR just adds the same functionality to the `yum` backend with the same interface.

I might be missing something and in that case, please explain why this case is different from the `zypper` one.

> Would it be better for pkg.info_installed support filtering the attributes

Not really, that would cover a different use case. The point of this PR is to avoid having to use `pkg.info_installed` after a `pkg.install` or `pkg.installed` to get all relevant information - we want it all directly from `pkg.install`/`pkg.installed`.

> Looks like the context_['pkg.list_pkgs'] will not work after this change.

I do not see why this PR should introduce a regression, but if you can demonstrate that I am of course happy to fix it.

**Comment by damon-atkins** [comment]
From PR https://github.com/saltstack/salt/pull/42310
> It adds a new optional parameter to list_pkg in the zypper module to return more data than the version (original reason is that for SUSE Manager integration we also need arch and install_date). Format is the same of existing method info_installed.

This PR includes changes to `pkg.install` which I did not notice in the previous PR 43210 not mention in the PR description.  I was mainly pushing for consistency than anything else in the discussion

This extra information is needed by SUSE Manager which directly calls `pkg.install`?
As `pkg.list_pkgs` need to be updated to provide extra attributes beyond just version as it feeds data into `pkg.install`.   `pkg.install` needs to be modified to accept and request extra attributes, so that extra data can outputted when action is taken so that SUSE Manager can use this information and report on it.

e.g. SUSE Mange needs out put like this which this PR and 43210 accomplishes.
```
        {'<package>': {
             'old': {
                 'version': '<old-version>',
                 'arch': '<old-arch>'},
 
             'new': {
                 'version': '<new-version>',
                 'arch': '<new-arch>'}}}
```
Does that about some it up? Just after clarity on the PR goals and why something already in place can not do this job.   I have no issues assume above  understanding is correct.

Salt can run in two modes fork or threaded. As I am not  sure how this change is going to be used, I will give an example.
Lets say `_context__['pkg.list_pkgs']` is filled with data with out any extra attributes and then another request comes in which wants extra attributes, it will be provided with the cache/_context_ results which does not have the attributes. Tell you if this is an issue or not I would need to dig deeper into the code. Someone from the Salt Team might be able to tell you.  You might need to add `_context__['pkg.list_pkgs_attr_used'] =attr` so you can tell if the data is valid/what is required in `_context__['pkg.list_pkgs']`

**Comment by moio** [comment]
> This extra information is needed by SUSE Manager

Correct.

> which directly calls pkg.install?

Almost correct: it can call `pkg.install` but also `pkg.installed`, `pkg.removed` or any other state which indirectly calls those. Since state call parameters are passed down to execution modules (like `pkg.install`), these PRs cover all cases that SUSE Manager is interested in.

> As pkg.list_pkgs need to be updated to provide extra attributes beyond just version as it feeds data into pkg.install. pkg.install needs to be modified to accept and request extra attributes, so that extra data can be collect by SUSE Manager which can use this information and report on it. [...]
> Just after clarity on the PR goals and why something already in place can not do this job. I have no issues assume above understanding is correct.

Correct.

Just a note: while it's certainly true that the SUSE Manager integration has been the trigger for this work, I also think it has general value, so that's the reason why the description was worded this way (namely: this is not a SUSE Manager specific hack we try to forcefully push to everyone).

> Lets say `__context__['pkg.list_pkgs']` is filled with data with out any extra attributes and then another request comes in which wants extra attributes, it will be provided with the cache/context results which does not have the attributes. To tell you if this is an issue or not I would need to dig deeper into the code. 

Here are saving the "full version" of the `list_pkgs` result in `__context__`, then "filtering" it depending on arguments and returning the "filtered copy". So if you specify eg. `attr='["version", "arch"]'` on a first call, all attributes will be saved in `__context__` including eg. `install_date`, not just the ones requested via arguments. Any subsequent call with different `attr` arguments (eg. `attr='all'`) will thus still be served from the context cache and return appropriate attributes because a different filtering will happen in that case.

See:

https://github.com/moio/salt/blob/c0281e9052d01114508ecc093480475fe61a07d4/salt/modules/yumpkg.py#L625-L626 (save in cache, "filter" and return)
https://github.com/saltstack/salt/pull/42718/files#diff-b47e2d68762ff7826be85833f119061fR311 (implements the "filtering")

I would appreciate a positive review from you if you feel convinced by the above explanations!

Thanks :innocent:

File: salt/utils/pkg/rpm.py [hunk 1]
```diff
@@ -59,16 +60,17 @@ def check_32(arch, osarch=None):
     return all(x in ARCHES_32 for x in (osarch, arch))
 
 
-def pkginfo(name, version, arch, repoid):
+def pkginfo(name, version, arch, repoid, install_date, install_date_time_t):
```
Comment (damon-atkins) [inline salt/utils/pkg/rpm.py]: Seems pkginfo is only used by
salt/salt/modules/yumpkg.py:                pkginfo = salt.utils.pkg.rpm.pkginfo(**cur)
Have you followed the code through to check the impact of add extra parms and returning extra data?

**Review by damon-atkins — APPROVED** [review]
Just double check rpminfo() change impact.  Without running the code, and not see any other potential issues.

File: salt/utils/pkg/rpm.py (hunk 1 above)
Comment (moio) [inline salt/utils/pkg/rpm.py]: Yes, I did take a look at the call tree and it only boils down to `yumpkg` basically in the places interested by this PR. So it should be safe FMPOV.

Thanks!

**Comment by garethgreenaway** [comment]
Go Go Jenkins!

**Review by terminalmage — APPROVED** [review]
:+1:

**Comment by rallytime** [comment]
@moio There are some lint errors and test failures related to this change. Can you take a look?

Lint: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/13487/violations/
- salt/modules/yumpkg.py
- salt/utils/pkg/rpm.py
- tests/unit/modules/test_yumpkg.py
- salt/modules/pkg_resource.py

Tests: https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-cent7-py3/870/
- integration.modules.test_pkg.PkgModuleTest.test_hold_unhold
- integration.modules.test_pkg.PkgModuleTest.test_install_remove

**Comment by moio** [comment]
Thanks @rallytime, I overlooked those. Now they are all fixed and rebased on top of latest develop - all should be set now.

**Pull request merged**